main/autogen/constant_hash.cc

```diff
@@ -0,0 +1,128 @@
+#include "main/autogen/constant_hash.h"
+#include "common/typecase.h"
+#include "core/hashing/hashing.h"
+#include "parser/Node.h"
+
+using namespace std;
+namespace sorbet::autogen {
+
+// See the comment in `main/autogen/constant_hash.h` for the intention
+// and high-level behavior of `constantHashNode`.
+//
+// A few operational notes about this function:
+//
+// - the reason it takes a parsed AST (rather than, say, a desugared
+// one) is that this means we can do less allocation and run the hash
+// as quickly as possible.
+//
+// - the reason it takes a reference to a `unique_ptr` is so it
+// doesn't need to `move` or consume the AST: it just observes it. The
+// reason it takes a pointer rather than a reference is so we don't
+// need to check for nullability on every recursive call: we can check
+// once at the head of `hashNode` instead.
+//
+// - there's a specific weakness in the recursive implementation of
+// this function which I think is worth noting but shouldn't be a
+// problem in practice: in particular, if we have a bare constant
+// reference directly in the body of a class or module, that'll count
+// towards the hash, even though it doesn't change the set of
+// constants being defined or the inheritance thereof. That could be
+// fixed, but it seemed somewhat more complex, so I left the
+// implementation simpler. This should rarely be a problem in
+// practice, since it's relatively rare to find a bare constant
+// reference directly in that position.
+unsigned int hashNode(core::GlobalState &gs, unsigned int hashSoFar, const unique_ptr<parser::Node> &what) {
+    if (what.get() == nullptr) {
+        return hashSoFar;
+    }
+    typecase(
+        what.get(),
+        [&](parser::Const *cnst) {
+            // hash the scope then the name
+            hashSoFar = hashNode(gs, hashSoFar, cnst->scope);
+            hashSoFar = core::mix(hashSoFar, core::_hash("::"));
+            hashSoFar = core::mix(hashSoFar, core::_hash(cnst->name.shortName(gs)));
+        },
+        [&](parser::Send *send) {
+            // only need to handle if this is a `require`, `include`,
```

Comment:
Do we also need to handle `require_relative`?

Reply:
Luckily, no! The autoloader only cares about `require`; `require_relative` gets stripped out there.